README.md
# Blockscan Tags

Checks Curate for contract tags and posts them to an endpoint. Works nicely with cron.

> Note: At the time of publication, there is no endpoint that lets us check if a given tag was already posted, so we keep a local DB to avoid posting the same tag again. Losing this database means the tool will submit everything again.

## Volta

This project uses `volta.sh` for node version management. See `package.json` for the version.

## Running

Duplicate `.env.example` and fill the env variables. You can use network specific files such as `.env.kovan` and then `yarn start:kovan`.

## Cron Job

This script plays well with cron. Here is a suggetsion on how to use it with `volta.sh`.

1- In the project folder, `vim script-mainnet.sh` paste the following and save.
2- Paste the following:

```
#!/bin/bash

cd /home/ubuntu/blockscan-tags

PATH=/home/ubuntu/.volta/bin:$PATH

echo "Using volta version $(volta --version)"
echo "Using node version $(node --version)"
yarn start:mainnet
```

3- Give script execution rights `chmod +x script-mainnet.sh`.
4- Setup the cron job: `crontab -e` and add `0 * * * * cd ~/blockscan-tags && ./script-mainnet.sh >> output-mainnet.log 2>&1`

## Debugging

If using vscode, use the following to debug.

.vscode/launch.json:

```
{
  "version": "0.2.0",
  "configurations": [
    {
      "type": "node",
      "request": "attach",
      "name": "Typescript Server",
      "protocol": "inspector",
      "port": 9229,
      "restart": true,
      "localRoot": "${workspaceFolder}",
      "remoteRoot": "."
    }
  ]
}

```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## 1.1.0 (2022-03-09)


### Features

* add checks for duplicate items ([3c68027](https://github.com/kleros/blockscan-tags/commit/3c68027d408bef4a3162619a4e4048b9531a4a53))
* add missing validation rules ([caea883](https://github.com/kleros/blockscan-tags/commit/caea883af42921cdc6559f4d84259a8a3215a668))
* initial release ([f740e34](https://github.com/kleros/blockscan-tags/commit/f740e34c379c27caefac1dc43f52424228c0b5d1))
* migrate db ([69637ea](https://github.com/kleros/blockscan-tags/commit/69637ea2dbdcf7222ba326e4cec36590fbb5fbbc))
* split scripts ([4478716](https://github.com/kleros/blockscan-tags/commit/44787166f3ae6e82c8c2ce39e295f5a1a0f457e5))
* support for xdai ([8052d60](https://github.com/kleros/blockscan-tags/commit/8052d60ff4fde0f36cc6e9c4d24436e98d245bd7))
* update publicNameTag length from 20 to 33 ([217b57d](https://github.com/kleros/blockscan-tags/commit/217b57de7bb79ff9ce5a134e25342263399a369d))


### Bug Fixes

* add missing field ([4246041](https://github.com/kleros/blockscan-tags/commit/4246041a85cfe1c996ace1d9b08a03212da294c4))
* bump tag to 35 chars ([4002543](https://github.com/kleros/blockscan-tags/commit/4002543a8f8fba0cef00fa2aa392796893a5a01b))
* compare and write lower case public name tag ([61c3e92](https://github.com/kleros/blockscan-tags/commit/61c3e920086384600c8d0de3ffdfa1d97001b29d))
* don't send requests in parallel and add missing separator ([fd1e95d](https://github.com/kleros/blockscan-tags/commit/fd1e95d795b82056f771c52e9e789e52dc0e755c))
* don't use null parameters ([e9ded26](https://github.com/kleros/blockscan-tags/commit/e9ded26c0dbb587bde9dc9f307ee3ee3343bbd0f))
* incorrect query format and duplicate filter ([d1b6468](https://github.com/kleros/blockscan-tags/commit/d1b6468b82da71922e51b6dce166f5abb7df40c2))
* loop exiting too early ([4776ec7](https://github.com/kleros/blockscan-tags/commit/4776ec76c31dcb61e4ddc070aa0a9fd3b3e18f50))
* query URL ([a66156b](https://github.com/kleros/blockscan-tags/commit/a66156b5e71b416f395efad23c131170f3689d08))
* use volta for version management ([c12f045](https://github.com/kleros/blockscan-tags/commit/c12f0450b1f29e863f9664651e38f959f0eb279d))


src/dbMigrator.ts
/**
 * Will be ran to migrate from the previous non-dupe format to the new one
 * Get all addresses in lowerCase to make sure there are no dupe addresses
 * Get all nameTags as well to avoid dupe name tags
 * This assumes that never a tag passed and then was removed
 */

import { open } from 'lmdb' // or require
import conf from './config'
import tagsFromEndpoint from './tagsFromEndpoint'
import { Tag } from './types'

const blockscanDB = open({
  path: `db-1`,
})

const nameTagDB = open({
  path: `db-tag-1`,
})

const migrateDb = async () => {
  let registeredTags: Tag[] = []
  try {
    console.info(`Fetching registered tags from mainnet subgraph...`)
    const mainnetTags = await tagsFromEndpoint(
      conf.MAINNET_GTCR_SUBGRAPH_URL,
      conf.MAINNET_LIST_ADDRESS,
    )
    registeredTags = registeredTags.concat(mainnetTags)
    console.info(`Got ${registeredTags.length} registered tags.`)
  } catch (error) {
    console.error(`Failed to fetch items`, error)
    return
  }

  for (const tag of registeredTags) {
    const { props } = tag
    const [addressObj, publicNameTagObj] = props
    const address = addressObj.value
    let publicNameTag = publicNameTagObj.value || ``

    publicNameTag = publicNameTag.slice(0, 35)

    if (!/^0x[a-fA-F0-9]{40}$/.test(address)) continue // Not an ETH address.
    if (publicNameTag.length > 35) continue // Exceeds max length.
    // if (publicNote.length === 0) continue // Usually mandatory, but
    // in the first list some addresses were submitted without it.

    console.info('putting in db:', { address, publicNameTag })
    blockscanDB.putSync(address.toLowerCase(), true)
    nameTagDB.putSync(publicNameTag.toLowerCase(), true)
  }
}

migrateDb()


src/tagsFromEndpoint.ts
import { Tag } from './types'
import fetch from 'node-fetch'

const tagsFromEndpoint = async (
  subgraphEndpoint: string,
  list: string,
): Promise<Tag[]> => {
  const subgraphQuery = {
    query: `
      {
        registered: litems(where: { registryAddress: "${list.toLowerCase()}", status:  Registered }) {
          id
          props {
            value
          }
        }
        clearingRequested: litems(where: { registryAddress: "${list.toLowerCase()}", status:  ClearingRequested }) {
          id
          props {
            value
          }
        }
      }
    `,
  }

  const response = await fetch(subgraphEndpoint, {
    method: 'POST',
    body: JSON.stringify(subgraphQuery),
  })

  const { data } = await response.json()
  const { clearingRequested, registered } = data
  const registeredTags: Tag[] = clearingRequested.concat(registered)
  return registeredTags
}

export default tagsFromEndpoint


src/index.ts
import fetch from 'node-fetch'
import { open } from 'lmdb' // or require
import conf from './config'
import tagsFromEndpoint from './tagsFromEndpoint'
import { Tag } from './types'

const blockscanDB = open({ path: 'db-1' })
const nameTagDB = open({ path: 'db-tag-1' })

const init = async () => {
  let registeredTags: Tag[] = []
  try {
    console.info(`Fetching registered tags from mainnet subgraph...`)
    const mainnetTags = await tagsFromEndpoint(
      conf.MAINNET_GTCR_SUBGRAPH_URL,
      conf.MAINNET_LIST_ADDRESS,
    )
    console.info(`Fetching registered tags from xdai subgraph...`)
    const xdaiTags = await tagsFromEndpoint(
      conf.XDAI_GTCR_SUBGRAPH_URL,
      conf.XDAI_LIST_ADDRESS,
    )
    registeredTags = registeredTags.concat(mainnetTags).concat(xdaiTags)
    console.info(`Got ${registeredTags.length} registered tags.`)
  } catch (error) {
    console.error(`Failed to fetch items`, error)
    return
  }

  for (const tag of registeredTags) {
    const { props } = tag
    const [addressObj, publicNameTagObj, publicNoteObj, websiteObj] = props
    const address = addressObj.value
    const website = websiteObj.value || ``
    const publicNote = publicNoteObj.value || ``
    let publicNameTag = publicNameTagObj.value || ``

    publicNameTag = publicNameTag.slice(0, 35)

    if (!/^0x[a-fA-F0-9]{40}$/.test(address)) continue // Not an ETH address.
    if (blockscanDB.get(address.toLowerCase())) continue // Address already posted to API.
    if (publicNameTag.length > 35) continue // Exceeds max length.
    if (publicNote.length === 0) continue // Mandatory field.
    if (nameTagDB.get(publicNameTag.toLowerCase())) continue // Exact name tag already posted.

    try {
      const query = `
          https://repaddr.blockscan.com/reportaddressapi?apikey=${conf.API_KEY}&address=${address}&chain=ETH&actiontype=1&customname=${publicNameTag}&comment=${publicNote}&infourl=${website}`
      if (conf.NODE_ENV === 'production') {
        const resp = await fetch(query)
        console.info(await resp.json())
        blockscanDB.putSync(address.toLowerCase(), true)
        nameTagDB.putSync(publicNameTag.toLowerCase(), true)
      } else {
        console.info(
          "development mode. address and nameTag saved would've been:",
        )
        console.info(query)
        console.info(address.toLowerCase())
        console.info(publicNameTag.toLowerCase())
      }
    } catch (error) {
      console.error(`Failed to post tag for ${address}.`, error)
    }
  }

  console.info(`Done processing tags.`)
}

init()
